src/server-bindings/server-binding.ts

```diff
@@ -0,0 +1,130 @@
+import crypto from 'crypto';
+import { Response, Router } from 'express';
+import { Scope } from '../models/connect';
+import { EventEmitter } from 'events';
+import { WebhookTriggers } from '../models/webhook';
+import { WebhookDelta } from '../models/webhook-notification';
+import AccessToken from '../models/access-token';
+import Nylas from '../nylas';
+import {
+  openWebhookTunnel,
+  OpenWebhookTunnelOptions,
+} from '../services/tunnel';
+
+type Middleware = Router;
+
+export enum ServerEvents {
+  TokenExchange = 'token-exchange',
+}
+
+export type ServerBindingOptions = {
+  defaultScopes: Scope[];
+  routePrefix?: string;
+  clientUri?: string;
+  useDevelopmentWebsocketEventListener?:
+    | boolean
+    | Partial<OpenWebhookTunnelOptions>;
+};
+
+export abstract class ServerBinding extends EventEmitter
+  implements ServerBindingOptions {
+  nylasClient: Nylas;
+  defaultScopes: Scope[];
+  routePrefix?: string;
+  clientUri?: string;
+  useDevelopmentWebsocketEventListener?:
+    | boolean
+    | Partial<OpenWebhookTunnelOptions>;
+
+  static DEFAULT_ROUTE_PREFIX = '/nylas';
+  static NYLAS_SIGNATURE_HEADER = 'x-nylas-signature';
+  private _untypedOn = this.on;
+  private _untypedEmit = this.emit;
+
+  protected constructor(nylasClient: Nylas, options: ServerBindingOptions) {
+    super();
+    this.nylasClient = nylasClient;
+    this.defaultScopes = options.defaultScopes;
+    this.routePrefix = options.routePrefix;
+    this.clientUri = options.clientUri;
+    this.useDevelopmentWebsocketEventListener =
+      options.useDevelopmentWebsocketEventListener;
+  }
+
+  abstract buildMiddleware(): Middleware;
+
+  // Taken from the best StackOverflow answer of all time https://stackoverflow.com/a/56228127
+  public on = <K extends WebhookTriggers | ServerEvents>(
+    event: K,
+    listener: (
+      payload: K extends WebhookTriggers
+        ? WebhookDelta
+        : { accessTokenObj: AccessToken; res: Response }
+    ) => void
+  ): this => this._untypedOn(event, listener);
+
+  public emit = <K extends WebhookTriggers | ServerEvents>(
+    event: K,
+    payload: K extends WebhookTriggers
+      ? WebhookDelta
+      : { accessTokenObj: AccessToken; res: Response }
+  ): boolean => this._untypedEmit(event, payload);
+
+  verifyWebhookSignature(
+    secret: string,
+    xNylasSignature: string,
+    rawBody: Buffer
+  ): boolean {
+    const digest = crypto
+      .createHmac('sha256', secret)
+      .update(rawBody)
+      .digest('hex');
+    return digest === xNylasSignature;
+  }
+
+  startDevelopmentWebsocket(): void {
```

Comment:
This is a bit confusing that it has to be called manually. Would recommend running this automatically on mount if `useDevelopmentWebsocketEventListener` is truthy

Reply:
We removed mount in favour of this becoming middleware. As a result though we've moved out the config to the function parameter.